src/bindings/python/tests/test_graph/test_core.py

```diff
@@ -391,3 +391,27 @@ def test_discrete_type_info():
     assert n1.get_type_info().name >= n3.get_type_info().name
     assert n3.get_type_info().name < n1.get_type_info().name
     assert n3.get_type_info().name <= n1.get_type_info().name
+
+
+def test_shape_negative_index():
+    shape = Shape([1, 2, 3, 4, 5])
+    assert shape[-1] == 5
+    assert shape[-3] == 3
+
+
+def test_partial_shape_negative_index():
+    shape = PartialShape([1, 2, 3, 4, 5])
+    assert shape[-1] == 5
+    assert shape[-3] == 3
+
+
+def test_shape_slicing_step():
+    shape = Shape([1, 2, 3, 4, 5])
+    assert list(shape[::2]) == [1, 3, 5]
+    assert list(shape[1::2]) == [2, 4]
+
+
+def test_partial_shape_slicing_step():
+    shape = PartialShape([1, 2, 3, 4, 5])
+    assert list(shape[::2]) == [1, 3, 5]
+    assert list(shape[1::2]) == [2, 4]
```

Comment:
I thought of something like this to make the code more readable:
```python
@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
def test_shape_negative_index(shape_type):
    shape = shape_type([1, 2, 3, 4, 5])
    assert shape[-1] == 5
    assert shape[-3] == 3
    assert shape[-5] == 1


@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
def test_shape_slicing_step(shape_type):
    shape = shape_type([1, 2, 3, 4, 5])
    assert list(shape[::2]) == [1, 3, 5]
    assert list(shape[1::2]) == [2, 4]
    assert list(shape[::-1]) == [5, 4, 3, 2, 1]
    assert list(shape[::-2]) == [5, 3, 1]
```

To avoid using two identical `@pytest.mark.parametrize` decorators, I could create a class, but I'm not sure if it will be better. Certainly, it would be better from a lines of code perspective since I would avoid rewriting two identical decorators, but on the other hand, creating a class solely for the purpose of saving lines of code is something I don't quite approve of. I'll leave an example of how I would implement it:

```python
@pytest.mark.parametrize("shape_type", [Shape, PartialShape])
class TestShape:

    def test_shape_negative_index(self, shape_type):
        shape = shape_type([1, 2, 3, 4, 5])
        assert shape[-1] == 5
        assert shape[-3] == 3
        assert shape[-5] == 1

    def test_shape_slicing_step(self, shape_type):
        shape = shape_type([1, 2, 3, 4, 5])
        assert list(shape[::2]) == [1, 3, 5]
        assert list(shape[1::2]) == [2, 4]
        assert list(shape[::-1]) == [5, 4, 3, 2, 1]
        assert list(shape[::-2]) == [5, 3, 1]
```
Which implementation would you recommend, @jiwaszki ?

Reply:
I don't think there's a reason to go as far as creating a separate test class, two identical one-line decorators are perfectly fine. :) Interesting idea though.